### ISSUE TITLE
Problems compile jitcode_lyap for large number of equations

### ISSUE THREAD

**Issue by erik-stormy** [neurophysik/jitcode#21]
For simplicity I used your code from here:

https://raw.githubusercontent.com/neurophysik/jitcode/master/examples/SW_of_Roesslers.py

and I changed it for jitcode_lyap:
```

#! /usr/bin/env python3
# -- coding: utf-8 --

# https://raw.githubusercontent.com/neurophysik/jitcode/master/examples/SW_of_Roesslers.py

def small_world_network(number_of_nodes, nearest_neighbours, rewiring_probability):
   n = number_of_nodes
   m = nearest_neighbours//2
   
   A = np.zeros( (n,n), dtype=bool )
   for i in range(n):
      for j in range(-m,m+1):
         A[i,(i+j)%n] = True
   
   # rewiring
   for i in range(n):
      for j in range(i):
         if A[i,j] and (np.random.random() < rewiring_probability):
            A[j,i] = A[i,j] = False
            while True:
               i_new,j_new = np.random.randint(0,n,2)
               if A[i_new,j_new] or i_new==j_new:
                  continue
               else:
                  A[j_new,i_new] = A[i_new,j_new] = True
                  break
   
   return A

# example-start - CHANGE
from jitcode import jitcode_lyap, y, t
import numpy as np
import symengine

# change of GCC to CLANG
import os
os.environ["CC"] = "clang"

# parameters
# ----------

# LYAP:
# N = 500 too long
N = 500

# N = 10 Ok
# N = 20 ok
# N = 100 after a while

# jitcode without lyap
# N=500 Ok

ω = np.random.uniform(0.8,1.0,N)
a = 0.165
b = 0.2
c = 10.0
k = 0.01

# get adjacency matrix of a small-world network
A = small_world_network(
   number_of_nodes = N,
   nearest_neighbours = 20,
   rewiring_probability = 0.1
   )

# generate differential equations
# -------------------------------

sum_z = symengine.Symbol("sum_z")
helpers = [( sum_z, sum( y(3*j+2) for j in range(N) ) )]

def f():
   for i in range(N):
      coupling_sum = sum( y(3*j)-y(3*i) for j in range(N) if A[i,j] )
      coupling_term = k * symengine.sin(t) * coupling_sum
      yield -ω[i] * y(3*i+1) - y(3*i+2) + coupling_term
      yield  ω[i] * y(3*i) + a*y(3*i+1)
      coupling_term_2 = k * (sum_z-N*y(3*i+2))
      yield b + y(3*i+2) * (y(3*i) - c) + coupling_term_2

# integrate
# ---------

initial_state = np.random.random(3*N)

# CHANGE
ODE = jitcode_lyap(f, helpers=helpers, n_lyap=3*N)

ODE.generate_f_C(simplify=False, do_cse=False, chunk_size=150)

ODE.save_compiled(destination="jitcode_test_ly_N500")

print('n', N)

print('end of program')
```

I use:
jitcdde==1.5.0
jitcode==1.4.0
jitcsde==1.4.0
jitcxde-common==1.5.0
symengine==0.5.0

on Ubuntu 18.04.

**Comment by Wrzlprmft** [comment]
I haven’t had the time to look deeper into this yet, but a few brief notes:

* Your versions of everything are not the latest.
* `n_lyap=3*N` means that you are computing 300 Lyapunov exponents. I am quite confident that you do not need those.

**Comment by Wrzlprmft** [comment]
I adjusted the number of Lyapunov exponents in your code:

``` python3
ODE = jitcode_lyap(f, helpers=helpers, n=3*N, n_lyap=1)
```

This runs in a few minutes without problems for me for `N=500`.

**Comment by erik-stormy** [comment]
Thank you so much for help. I actually did not know I can count less Lyapunov exponents than is order of equation!
I was able upgrade jitcode to this:
jitcdde==1.8.0
jitcode==1.6.0
jitcsde==1.6.0
jitcxde-common==1.5.2 
and also use: clang-9 and not just clang. Otherwise I tried N=500 and n_lyap=10 and just compilation was done after say 30 minutes and library has 32MB. I also tried higher n_lyap but it took more than 30 minutes and I stopped it. But I still did not tried to run actual computation.

**Comment by Wrzlprmft** [comment]
Once you successfully compiled, the actual computation should be no problem. That’s the entire point of the module.

Still, what do you need 10 Lyapunov exponents for?

**Comment by erik-stormy** [comment]
To be honest I am not sure how many Lyapunov exponents I need. It is quite experimental stuff. I do not use this network of oscillators, but rather simple 2D partial differential equation and I want to use its solution to compute Finite Size Lyapunov exponent with software called lagrangian. And spectrum of classical exponents could be compared with. So maybe will count some of them, or non of them. I suppose that jitcode_lyap counts q-highest exponents.
Basically, I counted 10 of them just to try how many exponents is possible in reasonable time - in this particular case it is not important at all. I think I chose Roessler example to be sure about correctness of the specified problem - to eliminate possible errors in code of my equation.

**Comment by Wrzlprmft** [comment]
I had a quick look at [Lagrangian](https://lagrangian.readthedocs.io/en/latest/contents.html#ftle-and-fsle-definition-and-properties) now, and if I understand correctly, it’s mainly about maximal Lyapunov exponents, but with a local separation. This is something that is already doable with `jitcode_restricted_lyap`, but it’s not the most efficient way of doing this kind of thing and it should be easy to implement a special function for this. Be aware that JiTCODE already provides finite-time Lyapunov exponents and vectors (in the sense of Lagrangian, IIUC).

If you are familiar with what exactly this kind of study needs, I am happy to discuss how to equip JiTCODE for this with you. From what I read, this does not seem like much work for me (as almost all of the ingredients already exist) and a useful addition.

> I suppose that jitcode_lyap counts q-highest exponents.

Correct. All numerical methods for Lyapunov exponents yield them in descending order.
